Mathlib/FieldTheory/Perfect.lean

```diff
@@ -0,0 +1,104 @@
+/-
+Copyright (c) 2023 Oliver Nash. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Oliver Nash
+-/
+import Mathlib.FieldTheory.Separable
+
+/-!
+
+# Perfect fields and rings
+
+In this file we define perfect fields, together with a generalisation to (commutative) rings in
+prime characteristic.
+
+## Main definitions / statements:
+ * `PerfectRing`: a ring of characteristic `p` (prime) is said to be perfect in the sense of Serre,
+   if its Frobenius map is bijective.
+ * `PerfectField`: a field `K` is said to be perfect if every irreducible polynomial over `K` is
+   separable.
+
+-/
+
+open Function
+
+/-- A perfect ring of characteristic `p` (prime) in the sense of Serre.
+
+NB: This is not related to the concept with the same name introduced by Bass (related to projective
+covers of modules). -/
+class PerfectRing (R : Type _) (p : ℕ) [CommRing R] [Fact p.Prime] [CharP R p] : Prop where
```

Comment:
Doesn't this clash with the existing `PerfectRing`?

Reply:
Yes, once I'm happy with this file, I'll be nuking the existing `PerfectRing`.